### ISSUE TITLE
globalReturn is not reset when node environment is set to false

### ISSUE THREAD

**Issue by ilyavolodin** [eslint/eslint#3915]
Using ESLint v1.5.0 and the following directory structure:

```
..
.eslintrc
src
 - .eslintrc
 - test.js
```

If the top level .eslintrc file has
```
{
  "env": { "node": true }
}
```
and the one inside `src` directory has
```
{
  "env": { "node": false, "browser": true }
}
```
For test.js file environments correctly cascade to `node: false` and `browser: true`, however, `globalReturn` ecmaFeature is still set to `true`. I don't think that's correct behavior. This can be reproduced by using `no-unused-vars`, declaring unused variable inside test.js and then adding `exported` comment with variable name. ESLint will report variable still being unused (will file another issue for this #3916). @mysticatea could you take a look? I think this is related to the fix for #3735 since this is behaving correctly in ESLint v1.4.0

**Comment by mysticatea** [comment]
Okay, I will do.

**Comment by mysticatea** [comment]
The cause is that [createEnvironmentConfig](https://github.com/eslint/eslint/blob/107f8be42917e2f0773aa0fabcc17c687dd30cdf/lib/config.js#L79-L108) checks only environments which are set `true`. I had mistaken checking inside of this function.

Now, `env`s have globals but we cannot turn globals off, so it cannot overwrite globals with `env: false`.
So I need to rewrite the overwriting-logic.

**Comment by ilyavolodin** [comment]
Sounds good. Thanks for looking into it.

**Comment by mysticatea** [comment]
It's hard.
I will try in a few days.

**Comment by nzakas** [comment]
I'll see if I can fix this

**Comment by rocketmonkeys** [comment]
@mysticatea @nzakas - I just ran into this bug and I can see where it's happening.  Are either of you currently working on a fix?  If not, I might try my hand (since I just spent an hour digging through this code).

**Comment by ilyavolodin** [comment]
@rocketmonkeys I would submit PR anyways (if it's not going to take a long time to do), since even if somebody else will work on it, they can verify their fix against yours. It's also been 4 days since @nzakas posted his message.

**Comment by nzakas** [comment]
I was going to look at it today, but feel free to submit a PR

**Comment by rocketmonkeys** [comment]
I have a fix, but not the tests; those are harder since I'm not nearly as familiar with the mocha/testing setup we have.  Can I just PR the fix as is, or do I need the test in there before I do anything?

**Comment by nzakas** [comment]
You can PR the fix if you'd like to get feedback, we just can't merge until there are tests.

**Comment by rocketmonkeys** [comment]
@nzakas - ok, prelim PR here: https://github.com/eslint/eslint/pull/4713

I'm guessing it's not complete, as this is my first dive into the code.  It works for my use-case (parent folder enables env, but subfolder disables it).  Not sure about others.  I'll work on the test, no idea when I'll have that done.

**Comment by nzakas** [comment]
I'm also going to dig into this.

**Comment by nzakas** [comment]
The root cause is that we are applying environment settings as files are being loaded, instead of waiting until all merging is complete. So basically, just have to wait until after merging to do that.

**Comment by gyandeeps** [comment]
Now that makes it sound simple :smile:

**Comment by mysticatea** [comment]
please remember this issue: https://github.com/eslint/eslint/issues/3735#issuecomment-141051376
